### PR TITLE
Fixes error when too many parameters are passed to fused cuda kernel

### PR DESCRIPTION
**Pull request by royju**
Bug fix for https://github.com/pytorch/pytorch/issues/15043, where a large fusion in JIT with a large number of kernel arguments, which exceeds the limit allowed by nvrtc on a cuda device.
  The fix is to check the number of arguments before a cuda kernel is generated. If the number exceeds the limit, take the runFallBack() path.
  Add a reduced test from the original issue to keep the test time low. The test would fail without this fix.

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -566,6 +566,45 @@ def __init__(self):
 
 
 class TestJit(JitTestCase):
+    @unittest.skipIf(not RUN_CUDA, "requires CUDA")
+    def test_large_nbr_kernel_args(self):
```
Comment (apaszke) [inline test/test_jit.py]: This test will be very slow. Can we just prepare a graph that has ~200 ops that would normally get fused perfectly, but we can't because that's too much for a single kernel?

**Review by apaszke — DISMISSED** [review]
We should avoid generating fusion groups this large altogether. This patch fixes the bug but silently disables an optimization in a case which is legitimate.

File: test/test_jit.py (hunk 1 above)
Comment (royju) [reply test/test_jit.py]: Thanks for the comment. The number of kernel arguments and the number of ops in a fusion group could be related but not always so, and the cause of this bug is the former. While we may want to limit the number of ops in a fusion group, I consider that's a separate issue to address. In fact, in the original test from the issue, the offending kernel itself has only about 10 ops in the fusion group, but because of a FusedConcat, it has a large number of live-ins, which exceeds the limit. Hence, we need to constrain the # of arguments, not the # of ops, for this specific issue.
   It's true that the fix gives up fusion on the entire FusionGroup if the number of args is over the limit. One could argue to limit the FusionGroup by # of args in GraphFuser. In fact, I did prototype that. But the trade-off isn't always straightforward. First, since this limit is due to the cuda path, a change in GraphFuser could affect other devices. In addition, it's possible that if one stops a FusionGroup by estimating the number of possible arguments during fusion, further fusion could actually reduce the number of live-in/out arguments in the FusionGroup back below the limit. To keep this PR simple, I didn't continue that route. As I have learned, GraphFuser is sometime too aggressive such that we bail out to the fall back path in the Executor too often than we'd like to see. When we address the aggressiveness of GraphFusion (including constraining # of ops), we will reduce such possibility in giving up an entire FusionGroup by taking a preventive measure (as opposed to a reactive one).
   Lastly on reducing the test case, I largely cut down the test and testing time from the original test. I again tried more just now, but haven't made much progress. The challenge is that I need to maintain the number of arguments (> 130) and also make sure the GraphFuser fuses it. Currently, the body of each iteration is
                    b = input[i] * 2
                    output.append(b)
To keep each iteration generating one new arg to a later FusedConcat, I keep the indexing form "input[i]", which is translated to 3 ops (1 select op and 2 constants: axis and unrolled index). To trigger fusion, I need to add a pw op (one mul op). 4 ops * 130 unrolled iterations are more than a couple of hundred ops. This test does take more time than some of the really small tests, but it doesn't seem totally out of line and it's off on the non-cuda path. I haven't found a good way to further reduce the test, probably because still being new to PyT, but I'm open to any suggestion.

Comment (apaszke) [reply test/test_jit.py]: @royju I agree those are orthogonal issues, but my point is exactly what you said -- you can reproduce this issue by using a much simpler test case, so you should use that instead.

Comment (apaszke) [reply test/test_jit.py]: Ok, my bad. The test does look somewhat simple, but please don't call it GRU in this case.

Also, does your test actually even trigger the fuser? You never run the traced function, so the code probably doesn't even get compiled.

Finally, regarding the tradeoffs, I think those are relatively simple. I doubt there are other devices where fusing more than 128 arguments into a single kernel would be beneficial. Finally, emitting many kernels is always superior to emitting no kernels.

Comment (royju) [reply test/test_jit.py]: Right. It's no longer the original GRU in the issue. I had a clarifying comment, but I will rename the class to avoid confusion.

Line 607    traced_gru = torch.jit.trace(gru, (input)) calls jit trace, so compilation, including fuser, does get invoked.

Ok. I can re-introduce the logic to estimate and track the # of arguments during fusion and stop growing if the limit is exceeded. I'm at GTC much of the week, and shall get back on this later this week. Thanks.

**Comment by royju** [comment]
Revision done according to the previous comments. Rename the class name in the test. Estimate the # of kernel arguments during fusion, and bail out if the limit may be exceeded.

**Comment by royju** [comment]
ci/circleci: binary_linux_conda_2.7_cpu_build — Your tests failed on CircleCI

Mar 22 01:52:03 /opt/conda/conda-bld/pytorch-nightly-cpu_1553219249856/work/third_party/ideep/mkl-dnn/src/cpu/ref_rnn.cpp:979:50: error: ‘void cblas_sgemm_free(float*)’ is deprecated (declared at /opt/conda/conda-bld/pytorch-nightly-cpu_1553219249856/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/include/mkl_cblas.h:804) [-Werror=deprecated-declarations]
Mar 22 01:52:03                  cblas_sgemm_free(weights(i, j, k));
Mar 22 01:52:03                                                                     ^

The above test failure is not caused by this commit. It also failed to other devs, and it appears the pytorch/pytorch repo has this issue already.

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 2]
```diff
@@ -1108,6 +1129,9 @@ struct GraphFuser {
         }
         any_fused = true;
         auto maybe_group = tryFuse(fused_cat, input);
+        if( !maybe_group ) {
+          continue;
```
Comment (ngimel) [inline torch/csrc/jit/passes/graph_fuser.cpp]: at this point `any_fused` is already set to `true`, even if no concat fusion happens, which will throw off subsequent checks.

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 3]
```diff
@@ -1066,7 +1083,11 @@ struct GraphFuser {
       auto subgraph = producer->node()->g(attr::Subgraph);
       auto* node = subgraph->outputs().at(producer->offset())->node();
       return node->kind() != prim::FusedConcat &&
-          !containsGradSumToSize(producer->node());
+          !containsGradSumToSize(producer->node()) &&
+          ((before_check->inputs().size() + before_check->outputs().size() +
+            producer->node()->inputs().size() + producer->node()->outputs().size())
+           <= fusion_kernel_args_limit);
+
```
Comment (ngimel) [inline torch/csrc/jit/passes/graph_fuser.cpp]: This seems unnecessary as the same check will be performed a few lines later in `tryFuse`, also, if you are trying to fuse concat to something else than a FusionGroup, this function will return `true` (line 1092) which does not look right.

File: torch/csrc/jit/passes/graph_fuser.cpp (hunk 2 above)
Comment (royju) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Thanks. Moved down "any_fused = true;" after all early bail-out are done.

File: torch/csrc/jit/passes/graph_fuser.cpp (hunk 3 above)
Comment (royju) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Remove this check and let the one in tryFuse() take care of this.

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 4]
```diff
@@ -209,6 +213,12 @@ struct GraphFuser {
       return false;
     if (!node->is_constant(attr::dim))
       return false;
+
+    Node* list_construct = node->namedInput(attr::tensors)->node();
```
Comment (ngimel) [inline torch/csrc/jit/passes/graph_fuser.cpp]: `list_construct` is the same as `tensors_node` a few lines below, you can move `tensors_node` line here and reuse it.

**Review by ngimel — CHANGES_REQUESTED** [review]

File: test/test_jit.py [hunk 5]
```diff
@@ -567,6 +567,42 @@ def __init__(self):
...
+        class Recurrence(nn.Module):
+            def __init__(self, input_size, seq_len):
+                super(Recurrence, self).__init__()
+                self.input_size = input_size
+                self.batch_first = True
```
[the remark this answers is not in the document]
Comment (royju) [reply test/test_jit.py]: Done.

File: test/test_jit.py [hunk 6]
```diff
@@ -567,6 +567,42 @@ def __init__(self):
...
+                self.seq_len = seq_len
+
+            def forward(self, input):
+                if self.batch_first:
+                    input = input.transpose(0, 1)
+
+                # Main loop
+                output = []
+                for i in range(self.seq_len):
```
[the remark this answers is not in the document]
Comment (royju) [reply test/test_jit.py]: Using input.size(0) leads to the following warning. Since this is a test, it should be ok. Change done.
test_large_nbr_kernel_args (__main__.TestJit) ... test/test_jit.py:581: TracerWarning: Converting a tensor to a Python index might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  for i in range(input.size(0)):

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 7]
```diff
@@ -201,6 +201,10 @@ struct GraphFuser {
       //   but this is checked later
       return isFusable(node->inputs()[0]->node());
     }
+    if( (node->inputs().size() + node->outputs().size()) >
```
[the remark this answers is not in the document]
Comment (royju) [reply torch/csrc/jit/passes/graph_fuser.cpp]: This check is inside isFusableMap(), which is called by isFusable() and one other place in tryToMoveChunk(). In the latter case, there is no immediate call to tryFuse() inside the function. It looks we should keep the guard. No?

File: torch/csrc/jit/passes/graph_fuser.cpp (hunk 4 above)
Comment (royju) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Done. Will push a commit soon after testing. Thanks.

File: test/test_jit.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (royju) [reply test/test_jit.py]: Put seq_len back as a module attribute to avoid the warning.

File: torch/csrc/jit/passes/graph_fuser.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (ngimel) [reply torch/csrc/jit/passes/graph_fuser.cpp]: `tryToMoveChunk` is not making any fusions by itself, thus should not blacklist any nodes. Any changes to FusionGroup are caused by tryFuse, thus checking (or estimating, as the case may be) number of inputs/outputs only in tryFuse is better.

[the remark this answers is not in the document]
Comment (royju) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Ok. Thanks for checking. This check in isFusableMap() is removed. Just pushed a new commit. CI is starting.

**Review by ngimel — APPROVED** [review]

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 8]
```diff
@@ -1106,8 +1118,11 @@ struct GraphFuser {
         if (!canFuseWithConcat(input, fused_cat)) {
           continue;
         }
-        any_fused = true;
         auto maybe_group = tryFuse(fused_cat, input);
+        if( !maybe_group ) {
+          continue;
+        }
+        any_fused = true;
         AT_ASSERT(maybe_group && maybe_group == fused_cat);
```
Comment (apaszke) [inline torch/csrc/jit/passes/graph_fuser.cpp]: The first part of this assert is meaningless now

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 9]
```diff
@@ -209,7 +209,12 @@ struct GraphFuser {
       return false;
     if (!node->is_constant(attr::dim))
       return false;
+
     auto tensors_node = node->namedInput(attr::tensors)->node();
+    if( (tensors_node->inputs().size() + node->outputs().size()) >
```
Comment (apaszke) [inline torch/csrc/jit/passes/graph_fuser.cpp]: FWIW there's always only a single output in here, so we don't have to account for that

File: torch/csrc/jit/passes/graph_fuser.cpp (hunk 2 above)
Comment (apaszke) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Please adjust `canFuseWithConcat` instead. The point here was to make sure that this function is sufficient for `tryFuse` to succeed, which is checked by the assert below.

**Review by apaszke — CHANGES_REQUESTED** [review]

File: torch/csrc/jit/passes/graph_fuser.cpp (hunk 8 above)
Comment (royju) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Ok. Revision is done in the latest commit. Place a limit check in canFuseWithConcat() to restore the original code sequence around tryFuse() call in fuseConcats().

File: torch/csrc/jit/passes/graph_fuser.cpp (hunk 9 above)
Comment (royju) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Since an output could still be an argument to the fused kernel generated later, I didn't change this part in the latest commit.

Comment (apaszke) [reply torch/csrc/jit/passes/graph_fuser.cpp]: BTW we should assert that this never happens, because now we have guards for it. If it does, then someone messed up somewhere.

Comment (royju) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Would you suggest a specific place and assert to add?

Comment (apaszke) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Right here. This is the condition that should never happen in this very place

**Review by apaszke — APPROVED** [review]
Please just change the condition inside the fuser (not the fuser pass) to be a hard assertion.

File: torch/csrc/jit/fuser/compiler.cpp [hunk 10]
```diff
@@ -329,8 +330,12 @@ std::shared_ptr<FusedKernel> compileKernel(
     }
   }
 
-  const std::string name = "kernel_" + std::to_string(next_kernel_id++);
+  // Have checked the limit at graph_fuser. Assert nothing else changing that.
+  AT_ASSERT((flat_inputs.size() + flat_outputs.size()) <=
+            fusion_kernel_args_limit);
+
```
Comment (ngimel) [inline torch/csrc/jit/fuser/compiler.cpp]: Now that this has come full circle, and compileKernel can no longer return nullopt, can you please change return type of this function, and remove nullopt handling logic in the executor?

Comment (royju) [reply torch/csrc/jit/fuser/compiler.cpp]: I was thinking to keep the more flexible interface able to handle the case of nullopt. But I can remove that if so desired.

Comment (royju) [reply torch/csrc/jit/fuser/compiler.cpp]: The limit check in compileKernel() has been turned to an assert, and the returned value of possible nullopt from compileKernel() has also been removed in the latest commits.
